app/controllers.rb

```diff
@@ -37,27 +37,62 @@
     data.to_json
   end
 
-  get :index, with: :endpoint, provides: [:json, :csv] do
-    options = get_search_args_from_params(params)
-    endpoint = options[:endpoint]
-    content_type options[:format].to_sym if options[:format]
-    DataMagic.logger.debug "-----> APP GET #{params.inspect}"
-
-    unless DataMagic.config.api_endpoints.keys.include? endpoint
-      halt 404, {
-        error: 404,
-        message: "#{endpoint} not found. Available endpoints: #{DataMagic.config.api_endpoints.keys.join(',')}"
-      }.to_json
-    end
+  get :index, with: ':endpoint/:command', provides: [:json] do
+    process_params
+  end
 
-    data = DataMagic.search(params, options)
-    halt 400, data.to_json if data.key?(:errors)
+  get :index, with: ':endpoint', provides: [:json, :csv] do
+    process_params
+  end
+end
 
-    if content_type == :csv
-      output_data_as_csv(data['results'])
-    else
-      data.to_json
-    end
+def process_params
+  options = get_search_args_from_params(params)
+  DataMagic.logger.debug "-----> APP GET #{params.inspect} with options #{options.inspect}"
+
+  check_endpoint!(options)
+  if required_params_present?(options)
+    # Only hit ES when the request is well formed.
+    set_content_type(options)
+    search_or_halt(options)
+  else
+    halt 400, {error: 400, message: 'Invalid parameters supplied'}.to_json
+  end
+end
+
+def required_params_present?(options)
+  if options[:command].nil? || options[:command] == 'stats' && options[:fields].present?
```

Comment:
Given that the `if` clause returns a boolean anyway, you can just have it be the body of the method without the `if`/`else`.